### PR TITLE
rustc_metadata: Merge items from `extern` blocks into their parent modules during metadata encoding rather than during metadata decoding

### PR DESCRIPTION
**Pull request by petrochenkov**

### CODE REVIEW

**Comment by petrochenkov** [comment]
We are doing less (in the worst case no more) work during both encoding and decoding, but apparently the `tcx.hir().item(*item_id)` part is somewhat expensive when called for all items in modules and eats the difference.

My plan was to simplify `fn each_child_of_item` until it can return an iterator, to avoid allocations in its users.
So I'm still interested in shifting this work from decoding to encoding somehow.

**Comment by Mark-Simulacrum** [comment]
It may be worth re-running now that https://github.com/rust-lang/rustc-perf/issues/1126 has at least partially been patched.

**Comment by petrochenkov** [comment]
Yeah, given that results in e.g. https://github.com/rust-lang/rust/pull/92156 don't make any sense I can only assume that it's all noise.

@bors try @rust-timer queue

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 1]
```diff
@@ -1100,9 +1100,21 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
...
         } else {
-            record!(self.tables.children[def_id] <- md.item_ids.iter().map(|item_id| {
-                item_id.def_id.local_def_index
-            }));
+            let direct_children = md.item_ids.iter().map(|item_id| item_id.def_id.local_def_index);
+            // Foreign items are planted into their parent modules from name resolution point of view.
+            let tcx = self.tcx;
+            let foreign_item_children = md
+                .item_ids
+                .iter()
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1108]: Would pre-filtering by `tcx.def_kind(item_id.def_id) == DefKind::ForeignMod` help?

**Comment by petrochenkov** [comment]
I entirely misinterpreted the regression source, it turns out that `try_item_ident` will format and allocate strings on its error path, so it's expensive to ask `if let Ok(ident) = self.try_item_ident(...)` on all items during decoding.

**Comment by petrochenkov** [comment]
>Small regression in instruction counts (up to 1.2% on incr-unchanged builds of deeply-nested-closures)

The results don't make any sense to me.
They are entirely different from the previous run (https://github.com/rust-lang/rust/pull/92153#issuecomment-1001997963) and local testing with `cg_diff` shows mostly unrelated changes in inlining choices around `fn imported_source_files`.
```
9,003,249 (9805.2%)  :<rustc_span::SourceFile as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
-8,972,806 (-9772.1%)  :<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::SourceFile], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::imported_source_files::{closure#3}::{closure#0}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<rustc_metadata::rmeta::decoder::ImportedSourceFile, <alloc::vec::Vec<rustc_metadata::rmeta::decoder::ImportedSourceFile> as alloc::vec::spec_extend::SpecExtend<rustc_metadata::rmeta::decoder::ImportedSourceFile, core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::SourceFile], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::imported_source_files::{closure#3}::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
   109,843 (119.6%)  :<rustc_span::span_encoding::Span as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
   -60,555 (-65.9%)  :<rustc_span::span_encoding::Span>::new
   -57,792 (-62.9%)  :<rustc_metadata::creader::CrateMetadataRef>::imported_source_files
    56,690 (61.74%)  :__memcpy_avx_unaligned_erms
    17,728 (19.31%)  :<alloc::borrow::Cow<str>>::into_owned
    -4,750 (-5.17%)  :<rustc_span::symbol::Symbol as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
    -4,452 (-4.85%)  :<(rustc_span::symbol::Symbol, rustc_span::def_id::DefIndex) as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
    -4,168 (-4.54%)  :<rustc_metadata::creader::CStore>::item_children_untracked
     3,800 ( 4.14%)  :<rustc_metadata::rmeta::decoder::DecodeContext as rustc_serialize::serialize::Decoder>::read_str
     3,520 ( 3.83%)  :<rustc_ast::ast::AttrStyle as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
     2,790 ( 3.04%)  :<rustc_ast::token::Lit as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
    -2,745 (-2.99%)  :<<rustc_ast::token::Lit as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode as core::ops::function::FnOnce<(&mut rustc_metadata::rmeta::decoder::DecodeContext,)>>::call_once
     2,604 ( 2.84%)  :<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[(rustc_span::symbol::Symbol, rustc_span::def_id::DefIndex)], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_diagnostic_items::{closure#0}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<(rustc_span::symbol::Symbol, rustc_span::def_id::DefId), <hashbrown::map::HashMap<rustc_span::symbol::Symbol, rustc_span::def_id::DefId, core::hash::BuildHasherDefault<rustc_hash::FxHasher>> as core::iter::traits::collect::Extend<(rustc_span::symbol::Symbol, rustc_span::def_id::DefId)>>::extend<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[(rustc_span::symbol::Symbol, rustc_span::def_id::DefIndex)], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_diagnostic_items::{closure#0}>>::{closure#0}>::{closure#0}>
     2,560 ( 2.79%)  :<rustc_ast::token::CommentKind as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
     1,428 ( 1.56%)  :llvm::PassRegistry::registerPass(llvm::PassInfo const&, bool)
    -1,370 (-1.49%)  :<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::def_id::DefIndex], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_struct_field_names::{closure#0}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<rustc_span::source_map::Spanned<rustc_span::symbol::Symbol>, <alloc::vec::Vec<rustc_span::source_map::Spanned<rustc_span::symbol::Symbol>> as alloc::vec::spec_extend::SpecExtend<rustc_span::source_map::Spanned<rustc_span::symbol::Symbol>, core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::def_id::DefIndex], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_struct_field_names::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
    -1,196 (-1.30%)  :__vfscanf_internal
    -1,171 (-1.28%)  :malloc_consolidate
      -910 (-0.99%)  :llvm::DenseMap<void const*, llvm::PassInfo const*, llvm::DenseMapInfo<void const*>, llvm::detail::DenseMapPair<void const*, llvm::PassInfo const*> >::grow(unsigned int)
      -773 (-0.84%)  :____strtoul_l_internal
       564 ( 0.61%)  :_int_free
      -535 (-0.58%)  :<rustc_ast::token::Token as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
       508 ( 0.55%)  :<rustc_span::source_map::SourceMap>::new_imported_source_file
       495 ( 0.54%)  :<rustc_metadata::rmeta::decoder::DecodeContext as rustc_serialize::serialize::Decoder>::read_enum_variant_arg::<rustc_ast::ast::Lit, <rustc_ast::ast::Lit as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode>
       495 ( 0.54%)  :<rustc_ast::ast::StrStyle as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
      -480 (-0.52%)  :<<rustc_ast::ast::StrStyle as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode as core::ops::function::FnOnce<(&mut rustc_metadata::rmeta::decoder::DecodeContext,)>>::call_once
       405 ( 0.44%)  :<rustc_metadata::rmeta::decoder::DecodeContext as rustc_middle::ty::codec::TyDecoder>::cached_ty_for_shorthand::<<&rustc_middle::ty::TyS as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode::{closure#0}>
      -275 (-0.30%)  :<rustc_metadata::rmeta::CrateDep as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
      -260 (-0.28%)  :<rustc_ast::ast::Attribute as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode
       241 ( 0.26%)  :<hashbrown::raw::RawTable<(rustc_span::source_map::StableSourceFileId, alloc::rc::Rc<rustc_span::SourceFile>)>>::insert::<hashbrown::map::make_hasher<rustc_span::source_map::StableSourceFileId, rustc_span::source_map::StableSourceFileId, alloc::rc::Rc<rustc_span::SourceFile>, core::hash::BuildHasherDefault<rustc_hash::FxHasher>>::{closure#0}>
       230 ( 0.25%)  :_dl_relocate_object
      -174 (-0.19%)  :<core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::def_id::DefIndex], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_struct_field_visibilities::{closure#0}> as core::iter::traits::iterator::Iterator>::fold::<(), core::iter::traits::iterator::Iterator::for_each::call<rustc_middle::ty::Visibility, <alloc::vec::Vec<rustc_middle::ty::Visibility> as alloc::vec::spec_extend::SpecExtend<rustc_middle::ty::Visibility, core::iter::adapters::map::Map<core::iter::adapters::map::Map<core::ops::range::Range<usize>, <rustc_metadata::rmeta::Lazy<[rustc_span::def_id::DefIndex], usize>>::decode<&rustc_metadata::creader::CrateMetadataRef>::{closure#0}>, <rustc_metadata::creader::CrateMetadataRef>::get_struct_field_visibilities::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
      -144 (-0.16%)  :<rustc_metadata::rmeta::decoder::DecodeContext as rustc_serialize::serialize::Decoder>::read_struct::<rustc_metadata::rmeta::CrateRoot, <rustc_metadata::rmeta::CrateRoot as rustc_serialize::serialize::Decodable<rustc_metadata::rmeta::decoder::DecodeContext>>::decode::{closure#0}>
      -140 (-0.15%)  :<rustc_metadata::creader::CStore>::get_crate_data
      -105 (-0.11%)  :memset
```

**Comment by pnkfelix** [comment]
This appears to be an improvement to the overall pipeline for compilation (assuming that in the common case overall, metadata is decoded far more often than encoded), and thus even if this small regression is real (which I'm not certain it is, especially given that all the data points are annotated with "?" in the perf output), I think we would accept the regression anyway.

@rustbot label: +perf-regression-triaged

**Pull request merged**